File: runtime/cpp/cyana-ast/PersistentAutomataAstApplication.cpp
//
// Created by tian wei jun on 2022/11/23 0023.
//

#include "PersistentAutomataAstApplication.h"
#include "Logger.h"
#include <iostream>
#include <list>

using namespace logger;

PersistentAutomataAstApplication::PersistentAutomataAstApplication() : persistentObject(0),
                                                                       dfaTokenAutomata(0),
                                                                       astAutomata(0) {
}
PersistentAutomataAstApplication::~PersistentAutomataAstApplication() {
  delete dfaTokenAutomata;
  dfaTokenAutomata = 0;
  delete astAutomata;
  astAutomata = 0;
  delete persistentObject;
  persistentObject = 0;
}

PersistentAutomataAstApplication::PersistentAutomataAstApplication(const std::string *persistentDataFilePath) {
  buildContext(persistentDataFilePath);
}

void PersistentAutomataAstApplication::buildContext(const std::string *persistentDataFilePath) {
  PersistentData *persistentData = new PersistentData(persistentDataFilePath);
  persistentObject = new PersistentObject(persistentData);
  dfaTokenAutomata = new DfaTokenAutomata(persistentObject->tokenDfa);
  astAutomata =
      new BacktrackingBottomUpAstAutomata(persistentObject->astDfa, persistentObject->startGrammar);
}

Ast *PersistentAutomataAstApplication::buildAst(const std::string *sourceCodeFilePath) {
  std::list<Token *> *tokens = dfaTokenAutomata->buildToken(sourceCodeFilePath);
  Ast *ast = astAutomata->buildAst(tokens);
  for (auto tokensIt = tokens->begin(); tokensIt != tokens->end(); tokensIt++) {
    Token *token = *tokensIt;
    std::cout << token->text << "  ";
    delete token;
    token = 0;
  }
  delete tokens;
  tokens = 0;
  Logger::info("CyanaAstApplication buildAst build successfully");
  return ast;
}

File: runtime/java/cyana-ast-runtime/src/cyana/ast/runtime/PersistentAutomataAstApplication.java
package cyana.ast.runtime;

import java.io.InputStream;
import java.util.List;

/**
 * 生成语法树的应用.
 *
 * @author tian wei jun
 */
public class PersistentAutomataAstApplication {
  PersistentObject persistentObject;
  DfaTokenAutomata dfaTokenAutomata;
  BacktrackingBottomUpAstAutomata astAutomata;

  public PersistentAutomataAstApplication(String persistentDataFilePath) {
    PersistentData persistentData = new PersistentData(persistentDataFilePath);
    this.buildContext(persistentData);
  }

  public PersistentAutomataAstApplication(InputStream inputStream) {
    PersistentData persistentData = new PersistentData(inputStream);
    this.buildContext(persistentData);
  }

  public void buildContext(PersistentData persistentData) {
    persistentObject = new PersistentObject(persistentData);
    dfaTokenAutomata = new DfaTokenAutomata(persistentObject.tokenDfa);
    astAutomata =
        new BacktrackingBottomUpAstAutomata(persistentObject.astDfa, persistentObject.startGrammar);
  }

  public Ast buildAst(String sourceCodeFilePath) {
    List<Token> tokens = dfaTokenAutomata.buildToken(sourceCodeFilePath);
    Ast ast = astAutomata.buildAst(tokens);
    Logger.info("CyanaAstApplication buildAst", "build successfully");
    return ast;
  }

  public Ast buildAst(InputStream inputStream) {
    List<Token> tokens = dfaTokenAutomata.buildToken(inputStream);
    Ast ast = astAutomata.buildAst(tokens);
    Logger.info("CyanaAstApplication buildAst", "build successfully");
    return ast;
  }
}
